src/main.rs

```diff
@@ -908,31 +908,67 @@ impl TypeDeroff {
 //
 //         return got_something
 
-// # Return whether the file at the given path is overwritable
-// # Raises IOError if it cannot be opened
-// def file_is_overwritable(path):
-//     result = False
-//     file = codecs.open(path, "r", encoding="utf-8")
-//     for line in file:
-//         # Skip leading empty lines
-//         line = line.strip()
-//         if not line:
-//             continue
-//
-//         # We look in the initial run of lines that start with #
-//         if not line.startswith('#'):
-//             break
-//
-//         # See if this contains the magic word
-//         if 'Autogenerated' in line:
-//             result = True
-//             break
-//
-//     file.close()
-//     return result
+#[test]
+fn test_file_is_overwritable() {
+    use std::path::Path;
+
+    // Create good test file
+    let good_string = String::from("good.txt");
+    let good_path = Path::new(&good_string);
+    tests::create_test_file(good_path);
+
+    // Create bad test path
+    let bad_string = String::from("bad.txt");
+    let bad_path = Path::new(&bad_string);
+
+    // Test for IO Error when File doesn't exist
+    let result = std::panic::catch_unwind(|| file_is_overwritable(bad_path));
+    assert!(result.is_err());
+
+    // Create bad test file
+    tests::create_test_file(bad_path);
+
+    // Tests
+    assert!(file_is_overwritable(good_path));
+    assert!(!file_is_overwritable(bad_path));
 
+    // Tear down
+    tests::remove_test_file(good_path);
+    tests::remove_test_file(bad_path);
+}
+
+// Return whether the file at the given path is overwritable
+// Raises IOError if it cannot be opened
 fn file_is_overwritable(path: &Path) -> bool {
-    unimplemented!()
+    use std::error::Error;
+    use std::fs::File;
+    use std::io::BufRead;
+    use std::io::BufReader;
+    let display = path.display();
+    let f = match File::open(path) {
+        Err(why) => panic!("couldn't open {}: {}", display, why.description()),
+        Ok(file) => file,
+    };
+    let file = BufReader::new(&f);
+    let mut result = false;
+    for line in file.lines() {
```

Comment:
Note, we'll want to use something like `split` here. Otherwise we'll encounter an otherwise resolvable panic if we read non-UTF8 bytes. Would also likely be worth it to pull in the `bstr` crate as a dependency so that we can use methods like [`trim`](https://docs.rs/bstr/0.1.2/bstr/struct.BStr.html#method.trim) on not-necessarily-UTF8 slices of bytes

Could iteraterify the loop as follows:
```rust
    Ok(file
        .split(b'\n')
        .map(|line| {
            // Okay to panic via `expect` here since we've already verified
            // that we can open the file for reading.
            bstr::B(&line.expect(&format!("I/O error encountered reading {}", display)))
                .trim()
                .to_owned()
        })
        .filter(|line| !line.is_empty())
        .take_while(|line| line.starts_with("#"))
        .any(|line| line.contains("Autogenerated")))
    // Note no need for separate `result`...we directly return the result of the iterator
```

Reply:
> // Okay to panic via `expect` here since we've already verified
> // that we can open the file for reading.

Why is the panic okay again?

Reply:
I could be wrong, but my interpretation was that the primary thing we'd need to guard against is a failure when initially opening the file and that subsequent IO errors would be unexpected enough that we don't want to build recovery logic around them. From the [documentation for split](https://doc.rust-lang.org/std/io/trait.BufRead.html#method.split):

> This function will yield errors whenever `read_until` would have also yielded an error.

Thoughts?

Reply:
Gotcha, yea that makes sense.